core/amber/src/main/python/core/runnables/test_batch_operator.py

```diff
@@ -0,0 +1,861 @@
+        )
+
+        # WorkerExecutionCompletedV2 should be triggered when workflow finishes
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                control_invocation=ControlInvocationV2(
+                    command_id=1,
+                    command=ControlCommandV2(
+                        worker_execution_completed=WorkerExecutionCompletedV2()
+                    ),
+                )
+            ),
+        )
+
+        # can process ReturnInvocation
+        input_queue.put(
+            ControlElement(
+                tag=mock_controller,
+                payload=set_one_of(
+                    ControlPayloadV2,
+                    ReturnInvocationV2(
+                        original_command_id=0, control_return=ControlReturnV2()
+                    ),
+                ),
+            )
+        )
+
+        reraise()
+
+    @pytest.mark.timeout(5)
+    def test_dp_thread_can_process_batch_complex(
+            self,
+            mock_udf,
+            mock_link,
+            mock_receiver_actor,
+            mock_controller,
+            input_queue,
+            output_queue,
+            mock_data_elements,
+            dp_thread,
+            mock_update_input_linking,
+            mock_add_partitioning,
+            mock_end_of_upstream,
+            mock_query_statistics,
+            mock_batch,
+            command_sequence,
+            reraise,
+    ):
+        dp_thread.start()
+
+        # can process UpdateInputLinking
+        input_queue.put(mock_update_input_linking)
+
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                return_invocation=ReturnInvocationV2(
+                    original_command_id=command_sequence,
+                    control_return=ControlReturnV2(),
+                )
+            ),
+        )
+
+        # can process AddPartitioning
+        input_queue.put(mock_add_partitioning)
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                return_invocation=ReturnInvocationV2(
+                    original_command_id=command_sequence,
+                    control_return=ControlReturnV2(),
+                )
+            ),
+        )
+
+        # can process a InputDataFrame
+        mock_udf.BATCH_SIZE = 10
+        for i in range(13):
+            input_queue.put(mock_data_elements[i])
+        time.sleep(0.1)
+
+        mock_udf.BATCH_SIZE = 20
+        for i in range(28):
+            input_queue.put(mock_data_elements[i])
+        time.sleep(0.1)
+
+        mock_udf.BATCH_SIZE = 5
+        for i in range(16):
+            input_queue.put(mock_data_elements[i])
+
+        output_data_element = []
+        for i in range(55):
+            output_data_element.append(output_queue.get())
```

Comment:
ditto.

Reply:
Without sleep with cause trouble on size change

Reply:
then it means the testing thread and dp thread are not synced properly. what trouble is it?

Reply:
It will cause the batch size not up to date.

Reply:
why is that? it sounds like an issue/bug. we should not make our test case adapt to code: test cases' purpose is to expose issue and bug.

Reply:
I think set batch size will cause a small delay. The whole process doesn't have any error, except the batch count is not correct.

Reply:
why would change of a batch size cause delay? please elaborate. 

> The whole process doesn't have any error, except the batch count is not correct.
The batch count is not correct **IS** the error.

Reply:
The batch count is not correct is because of the batch size change. "delay" is not a good description. For example, we changed batch size to 5 at the 41, but this processor process too fast and the batch size is changed to 5 even at the very beginning.

Reply:
try 
```suggestion
        mock_udf.BATCH_SIZE = 10
        for i in range(13):
            input_queue.put(mock_data_elements[i])
      
        output_data_element = []
        for i in range(10):
            output_data_element.append(output_queue.get())
            
            
        # change batch size after processing 13 tuples with batch size 10.
```

Reply:
May I ask what is "that's why we need to "sync" through output queue."? I think the batch size is determined during the input process.

Reply:
As discussed, you should let a thread wait for a condition to become true, instead of busy-waiting for some time arbitrarily.